R/read.R

```diff
@@ -54,3 +54,30 @@ function( descriptor, input, partial=FALSE){
         close(input)
 	message
 } )
+
+
+
+
+setGeneric( "readJSON", function( descriptor, input ){
+	standardGeneric( "readJSON" )
+} )
+
+setMethod( "readJSON", c( descriptor = "Descriptor" , input = "character" ),
+function(descriptor, input){
+	.Call( "Descriptor__readJSONFromString", descriptor@pointer, input,
+              PACKAGE = "RProtoBuf" )
+} )
+
+setMethod( "readJSON", c( descriptor = "Descriptor" ),
+function( descriptor, input){
+	if( !inherits( input, "connection" ) ){
+		stop( "can only read from connections" )
+	}
+	wasopen <- identical( summary(input)[["opened"]], "opened" )
+	if( !wasopen ) open( input, "rb" )
+        stopifnot(summary(input)[["text"]] == "binary")
+	message <- .Call( "Descriptor__readJSONFromConnection", descriptor@pointer, input,
+                         PACKAGE = "RProtoBuf" )
+	if ( !wasopen ) close(input)
```

Comment:
I noticed that in readASCII, you always close the connection. Let me know if you want me to do the same here.

Reply:
Probably. I think R catches when connections are open, and as their count is finite, it is best to give them back. Would be great if you could add that.

Reply:
Done.

Reply:
I really can't make up my mind which way is better.  In the same file, we do use the same `wasopen` logical you also set.  

Did you ever see warnings about connections left open either way?

Reply:
The answer to your question is not simple because it depends on more common usage patterns. I think checking `wasopen` is cleaner because it is clear about ownership of connections and is inline with what `readLines` does.

The main source of confusion is the usage of `readASCII(..., input = file(..., open = 'rb'))`, which is the pattern you follow in your unit tests. Using `file(..., open = 'rb')` as an argument will give you an open connection but with no handle to close it afterwards.

As the code is checked in right now, consider this pattern:
```
> con <- file(out.file)
> open(con, 'rb')
> tutorial.AddressBook$readASCII(con)
message of type 'tutorial.AddressBook' with 1 field set
> close(con)
Error in close.connection(con) : invalid connection
```

But it is not really a very big point, because I do not expect users to be passing open connections with the expectation to close them later themselves.

Reply:
Most likely I am not in the package and its code as much as you so if you can think of improvements on either side -- maybe even to `readASCII` ? -- don't hesitate to get in touch.